### PR TITLE
pattern matching

### PR DESCRIPTION
**Pull request by ggpwnkthx**
Refactored a bit to allow for regex pattern matching but can also be expanded on if necessary. The main idea here is that the process of getting the list of paths keys is separate method.

Example usage:
```
from aiopenapi3.extra import Reduced

class MSGraphReduced(Reduced):
    def __init__(self):
        super().__init__(
            {
                "/me/profile": None,
                re.compile(r"/me/mailFolders.*"): None,
                re.compile(r"/me/sendMail.*"): None,
            }
        )
```

### CODE REVIEW

File: aiopenapi3/extra.py [hunk 1]
```diff
@@ -1,40 +1,61 @@
...
+        for path in matched_paths:
+            paths[path] = self._extract_path_info(ctx, path)
+
+        ctx.document["paths"] = paths
         return ctx
 
+    def _find_matched_paths(self, ctx: "Document.Context") -> List[str]:
+        """Find paths in the context that match the given operations."""
+        return [
+            path
```
Comment (commonism) [inline aiopenapi3/extra.py:30]: this is the paths path - a str

File: aiopenapi3/extra.py [hunk 2]
```diff
@@ -1,40 +1,61 @@
...
+            for path_pattern in self.operations.keys()
+            for path in ctx.document["paths"].keys()
+            if (isinstance(path_pattern, str) and path_pattern == path)
+            or (isinstance(path_pattern, re.Pattern) and path_pattern.match(path))
+        ]
+
+    def _extract_path_info(self, ctx: "Document.Context", path: str) -> Dict:
+        """Extract path information from the context based on the given path."""
+        methods = self.operations[path]
```
Comment (commonism) [inline aiopenapi3/extra.py:39]: this tries to access a re match with the match instead of the re - does not work

**Comment by commonism** [comment]
rebased and tested with 

```diff
diff --git a/tests/extra_test.py b/tests/extra_test.py
index 017b7fe..36ea4be 100644
--- a/tests/extra_test.py
+++ b/tests/extra_test.py
@@ -1,4 +1,5 @@
 import sys
+import re
 
 if sys.version_info >= (3, 9):
     from pathlib import Path
@@ -102,7 +103,7 @@ def test_reduced(with_extra_reduced, httpx_mock):
         "http://127.0.0.1/api.yaml",
         with_extra_reduced,
         session_factory=httpx.Client,
-        plugins=[Reduced({"/B": None})],
+        plugins=[Reduced({re.compile("/B"): None})],
         loader=FileSystemLoader(Path("tests/fixtures")),
     )
     assert "/A/{Path}" not in api.paths.paths
```

does not work, the re has to be used to index access the operations, the path has to be used to index access the paths.
only the path is preserved.

**Comment by commonism** [comment]
Hrm, I just cleaned it up and fixed it … can you undo the removal of the branch?

**Comment by ggpwnkthx** [comment]
I apologies, I got trigger happy with the git cli and re-cloned your `reduced` branch.

**Comment by commonism** [comment]
I got a copy here including my change - merge into reduced?

**Pull request closed without merging**